src/AutoRest.CSharp/Common/Output/Models/Types/SignatureTypeProvider.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Reflection.Metadata;
+using System.Text;
+using System.Threading.Tasks;
+using AutoRest.CSharp.Generation.Types;
+using AutoRest.CSharp.Generation.Writers;
+using AutoRest.CSharp.Input.Source;
+using AutoRest.CSharp.Mgmt.AutoRest;
+using AutoRest.CSharp.Output.Models;
+using AutoRest.CSharp.Output.Models.Shared;
+using AutoRest.CSharp.Output.Models.Types;
+using Humanizer;
+using Microsoft.CodeAnalysis;
+using MethodParameter = AutoRest.CSharp.Output.Models.Shared.Parameter;
+
+namespace AutoRest.CSharp.Common.Output.Models.Types
+{
+    internal class SignatureTypeProvider : TypeProvider
+    {
+        private SourceInputModel? _sourceInputModel;
+
+        protected SignatureTypeProvider(string defaultNamespace, string defaultName, IEnumerable<MethodSignature> methods, SourceInputModel? sourceInputModel = null) : base(defaultNamespace, sourceInputModel)
+        {
+            DefaultName = defaultName;
+            DefaultAccessibility = "public";
+            Methods = methods;
+            _sourceInputModel = sourceInputModel;
+        }
+
+        private SignatureTypeProvider? _customization;
+        protected SignatureTypeProvider? Customization => _customization ??= CreateFromCompilation(_sourceInputModel?.Customization);
+
+        private SignatureTypeProvider? _previousContract;
+        protected SignatureTypeProvider? PreviousContract => _previousContract ??= CreateFromCompilation(_sourceInputModel?.PreviousContract);
+
+        // TODO: store the implementation of missing methods along with declaration
+        public virtual IEnumerable<MethodSignature> Methods { get; }
+
+        protected override string DefaultName { get; }
+
+        protected override string DefaultAccessibility { get; }
+
+        public void WriteMissingOverloadMethod(CodeWriter writer, MethodSignature currentMethodToCall, MethodSignature previousMethodToAdd, IList<MethodParameter> missingParameters)
```

Comment:
No sure if we should put a write method in a `TypeProvider`.

Reply:
The purpose it to reuse this logic in multiple writers, not only `ModelFactoryWriter`

Reply:
Moved to `BreakingChangeWriter`